shogun-lib/src/main/java/de/terrestris/shogun/lib/envers/CustomAnnotationRevisionMetadata.java

```diff
@@ -0,0 +1,162 @@
+ * limitations under the License.
+ */
+package de.terrestris.shogun.lib.envers;
+
+import org.springframework.data.history.RevisionMetadata;
+import org.springframework.data.util.AnnotationDetectionFieldCallback;
+import org.springframework.data.util.Lazy;
+import org.springframework.util.Assert;
+import org.springframework.util.ReflectionUtils;
+
+import java.lang.annotation.Annotation;
+import java.time.Instant;
+import java.time.LocalDateTime;
+import java.time.ZoneOffset;
+import java.util.Date;
+import java.util.HashSet;
+import java.util.Optional;
+import java.util.Set;
+
+/**
+ * A {@link RevisionMetadata} implementation that inspects the given object for fields with the configured annotations
+ * and returns the field's values on calls to {@link #getRevisionInstant()} and
+ * {@link #getRevisionNumber()}.
+ *
+ * @author Oliver Gierke
+ * @author Jens Schauder
+ */
+public class CustomAnnotationRevisionMetadata<N extends Number & Comparable<N>> implements RevisionMetadata<N> {
+
+    private final Object entity;
+    private final Lazy<Optional<N>> revisionNumber;
+    private final Lazy<Optional<Object>> revisionDate;
+    private final RevisionType revisionType;
+    private final Set<String> changedFields;
+
+    /**
+     * Creates a new {@link CustomAnnotationRevisionMetadata} inspecting the given entity for the given annotations. If no
+     * annotations will be provided these values will not be looked up from the entity and return {@literal null}. The
+     * revisionType will be set to {@literal unknown}
+     *
+     * @param entity must not be {@literal null}.
+     * @param revisionNumberAnnotation must not be {@literal null}.
+     * @param revisionTimeStampAnnotation must not be {@literal null}.
+     */
+    public CustomAnnotationRevisionMetadata(Object entity, Class<? extends Annotation> revisionNumberAnnotation,
+                                      Class<? extends Annotation> revisionTimeStampAnnotation) {
+
+        this(entity, revisionNumberAnnotation, revisionTimeStampAnnotation, RevisionType.UNKNOWN, new HashSet<>());
+    }
+
+    /**
+     * Creates a new {@link org.springframework.data.history.AnnotationRevisionMetadata} inspecting the given entity for the given annotations. If no
+     * annotations will be provided these values will not be looked up from the entity and return {@literal null}.
+     *
+     * @param entity must not be {@literal null}.
+     * @param revisionNumberAnnotation must not be {@literal null}.
+     * @param revisionTimeStampAnnotation must not be {@literal null}.
+     * @param revisionType must not be {@literal null}.
+     * @since 2.2.0
+     */
+    public CustomAnnotationRevisionMetadata(Object entity, Class<? extends Annotation> revisionNumberAnnotation,
+                                      Class<? extends Annotation> revisionTimeStampAnnotation, RevisionType revisionType, Set<String> changedFields) {
+
+        Assert.notNull(entity, "Entity must not be null!");
```

Comment:
Are assertions enabled in SHOGun or is this required / inherited from hibernate envers?

Reply:
yes this is from spring-data-envers

as this doesn't use the java `assert` keyword, but an import instead (Spring assertions) I don't think it needs to be enabled in jvm